spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepository.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.config.server.environment;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectIdBuilder;
+import org.yaml.snakeyaml.Yaml;
+
+import org.springframework.cloud.config.environment.Environment;
+import org.springframework.cloud.config.environment.PropertySource;
+import org.springframework.cloud.config.server.config.ConfigServerProperties;
+import org.springframework.core.Ordered;
+import org.springframework.util.StringUtils;
+
+/**
+ * @author Clay McCoy
+ */
+public class AwsS3EnvironmentRepository implements EnvironmentRepository, Ordered {
+
+	private final AmazonS3 s3Client;
+
+	private final String bucketName;
+
+	private final ConfigServerProperties serverProperties;
+
+	protected int order = Ordered.LOWEST_PRECEDENCE;
+
+	public AwsS3EnvironmentRepository(AmazonS3 s3Client, String bucketName,
+			ConfigServerProperties server) {
+		this.s3Client = s3Client;
+		this.bucketName = bucketName;
+		this.serverProperties = server;
+	}
+
+	@Override
+	public int getOrder() {
+		return this.order;
+	}
+
+	public void setOrder(int order) {
+		this.order = order;
+	}
+
+	@Override
+	public Environment findOne(String specifiedApplication, String specifiedProfile,
+			String specifiedLabel) {
+		final String application = StringUtils.isEmpty(specifiedApplication)
+				? serverProperties.getDefaultApplicationName() : specifiedApplication;
+		final String profile = StringUtils.isEmpty(specifiedProfile)
+				? serverProperties.getDefaultProfile() : specifiedProfile;
+		final String label = StringUtils.isEmpty(specifiedLabel)
+				? serverProperties.getDefaultLabel() : specifiedLabel;
+		StringBuilder objectKeyPrefix = new StringBuilder(application).append("-")
+				.append(profile);
+		if (!StringUtils.isEmpty(label)) {
+			objectKeyPrefix.append("-").append(label);
```

Comment:
I see, it does profile, then label. Interesting. In the `NativeEnvironmentRepository` label is used as a directory. Maybe we could do something similar rather than `-{label}`.

Reply:
There will always be a label, so it should find files without it as well.

Reply:
Yes, I wasn't sure how to best map label to s3 concepts. I saw that in Git it represented a branch. I briefly thought about it representing a version in an s3 bucket or the name of the bucket, but neither seemed right so I just put it at the end.
Label is optional so there won't always be a label right? I could use label as a prefix/directory. Those can be arbitrarily deep, I'm wondering if there is anything limiting label to be only one level deep (without path separators).

Reply:
I'd prefer the directory style structure since we do that in other places. One level is fine. Though one level in `SearchPathLocator` won't work.

Reply:
"Though one level in SearchPathLocator won't work."
Can you expand on that a bit?

Reply:
Right, so SearchPathLocator be somewhat moot when only one level is allowed? I guess it would return all files in that directory?

Reply:
fixed, label now denotes a directory path inside a bucket.